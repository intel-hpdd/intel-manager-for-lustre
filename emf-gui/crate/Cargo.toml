[package]
authors = ["EMF Team <emf@whamcloud.com>"]
categories = ["wasm", "rust", "tailwindcss", "seed"]
description = "The EMF GUI"
edition = "2018"
license = "MIT"
name = "emf-gui"
readme = "./README.md"
repository = "https://github.com/whamcloud/exascaler-management-framework"
version = "0.3.0"

[lib]
crate-type = ["cdylib"]

[dev-dependencies]
rand_core = "0.5.1"
rand_xoshiro = "0.4.0"
wasm-bindgen-test = "0.3"

[dependencies]
chrono = {version = "0.4", features = ["wasmbind"]}
chrono-humanize = "0.0.11"
futures = "0.3"
gloo-timers = {version = "0.2", features = ["futures"]}
humantime = "2.0"
im = {version = "15", features = ["serde"]}
emf-api-utils = {path = "../../emf-api-utils", version = "0.4"}
emf-graphql-queries = {version = "0.2", path = "../../emf-graphql-queries"}
emf-influx = {path = "../../emf-influx", version = "0.2"}
emf-wire-types = {path = "../../emf-wire-types", version = "0.4"}
js-sys = "0.3"
lazy_static = "1.4.0"
natord = "1.0"
number-formatter = {path = "../../number-formatter", version = "0.2"}
seed = "0.6.0"
serde = {version = "1", features = ['derive']}
serde_json = "1.0"
serde_urlencoded = "0.7.0"
wasm-bindgen = {version = "0.2", features = ["serde-serialize"]}
wasm-bindgen-futures = "0.4"

[dependencies.regex]
default-features = false
features = ["std", "unicode-perl"]
version = "1.3"

[dependencies.web-sys]
features = [
  "DomRect",
  "DomRectReadOnly",
  "Element",
  "EventSource",
  "HtmlDocument",
  "MessageEvent",
  "Navigator",
  "Notification",
  "NotificationOptions",
  "NotificationPermission",
  "ServiceWorkerContainer",
  "ServiceWorkerRegistration",
  "Window",
]
version = "^0.3"

[profile.release]
codegen-units = 1
lto = true
opt-level = 'z'
