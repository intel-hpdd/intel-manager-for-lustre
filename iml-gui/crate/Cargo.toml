[package]
authors = ["IML Team <iml@whamcloud.com>"]
categories = ["wasm", "rust", "tailwindcss", "seed"]
description = "The IML GUI"
edition = "2018"
license = "MIT"
name = "iml-gui"
readme = "./README.md"
repository = "https://github.com/whamcloud/integrated-manager-for-lustre"
version = "0.3.0"

[lib]
crate-type = ["cdylib"]

[dev-dependencies]
rand_core = "0.5.1"
rand_xoshiro = "0.4.0"
wasm-bindgen-test = "0.3"

[dependencies]
chrono = {version = "0.4", features = ["wasmbind"]}
chrono-humanize = "0.0.11"
futures = "0.3"
gloo-timers = {version = "0.2", features = ["futures"]}
humantime = "2.0"
im = {version = "15", features = ["serde"]}
iml-api-utils = {path = "../../iml-api-utils", version = "0.4"}
iml-graphql-queries = {version = "0.2", path = "../../iml-graphql-queries"}
iml-influx = {path = "../../iml-influx", version = "0.2"}
iml-wire-types = {path = "../../iml-wire-types", version = "0.4"}
js-sys = "0.3"
lazy_static = "1.4.0"
natord = "1.0"
number-formatter = {path = "../../number-formatter", version = "0.2"}
seed = "0.6.0"
serde = {version = "1", features = ['derive']}
serde_json = "1.0"
serde_urlencoded = "0.7.0"
wasm-bindgen = {version = "0.2", features = ["serde-serialize"]}
wasm-bindgen-futures = "0.4"

[dependencies.regex]
default-features = false
features = ["std", "unicode-perl"]
version = "1.3"

[dependencies.web-sys]
features = [
  "DomRect",
  "DomRectReadOnly",
  "Element",
  "EventSource",
  "HtmlDocument",
  "MessageEvent",
  "Navigator",
  "Notification",
  "NotificationOptions",
  "NotificationPermission",
  "ServiceWorkerContainer",
  "ServiceWorkerRegistration",
  "Window",
]
version = "^0.3"

[profile.release]
codegen-units = 1
lto = true
opt-level = 'z'
