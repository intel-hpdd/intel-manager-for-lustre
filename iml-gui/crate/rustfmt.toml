# https://github.com/rust-lang/rustfmt/blob/master/Configurations.md

edition = "2018"
newline_style = "Unix"
use_field_init_shorthand = true
use_try_shorthand = true

max_width = 120
