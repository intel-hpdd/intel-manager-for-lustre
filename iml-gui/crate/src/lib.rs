// @TODO: uncomment once https://github.com/rust-lang/rust/issues/54726 stable
//#![rustfmt::skip::macros(class)]

#![allow(clippy::used_underscore_binding)]
#![allow(clippy::non_ascii_literal)]
#![allow(clippy::enum_glob_use)]

mod generated;
mod page;

use generated::css_classes::C;
use iml_wire_types::warp_drive;
use js_sys::Function;
use seed::{events::Listener, prelude::*, *};
use std::{cmp, collections::HashMap, mem};
use wasm_bindgen::JsCast;
use web_sys::{EventSource, MessageEvent};
use Visibility::*;

const TITLE_SUFFIX: &str = "IML";
const USER_AGENT_FOR_PRERENDERING: &str = "ReactSnap";
const STATIC_PATH: &str = "static";
const SLIDER_WIDTH_PX: u32 = 5;
const MAX_SIDE_PERCENTAGE: f32 = 35f32;

#[derive(Clone, Copy, Eq, PartialEq)]
pub enum Visibility {
    Visible,
    Hidden,
}

impl Visibility {
    pub fn toggle(&mut self) {
        *self = match self {
            Visible => Hidden,
            Hidden => Visible,
        }
    }
}

#[derive(Debug, Copy, Clone)]
pub enum WatchState {
    Watching,
    Open,
    Close,
}

impl Default for WatchState {
    fn default() -> Self {
        WatchState::Close
    }
}

impl WatchState {
    pub fn is_open(self) -> bool {
        match self {
            WatchState::Open => true,
            _ => false,
        }
    }
    pub fn is_watching(self) -> bool {
        match self {
            WatchState::Watching => true,
            _ => false,
        }
    }
    pub fn should_update(self) -> bool {
        self.is_watching() || self.is_open()
    }
    pub fn update(&mut self) {
        match self {
            WatchState::Close => {
                mem::replace(self, WatchState::Watching);
            }
            WatchState::Watching => {
                mem::replace(self, WatchState::Open);
            }
            WatchState::Open => {
                mem::replace(self, WatchState::Close);
            }
        }
    }
}

// ------ ------
//     Model
// ------ ------

pub struct Model {
    pub page: Page,
    pub menu_visibility: Visibility,
    pub in_prerendering: bool,
    pub config_menu_state: WatchState,
    pub track_slider: bool,
    pub side_width_percentage: f32,
    pub records: warp_drive::Cache,
    pub locks: warp_drive::Locks,
}

#[derive(Clone, Copy, Eq, PartialEq)]
pub enum Page {
    Dashboard,
    Home,
    About,
    NotFound,
    Jobstats,
}

impl Page {
    pub fn to_href(self) -> &'static str {
        match self {
            Self::Home => "/",
            Self::Dashboard => "/dashboard",
            Self::Jobstats => "/jobstats",
            Self::About => "/about",
            Self::NotFound => "/404",
        }
    }
}

impl ToString for Page {
    fn to_string(&self) -> String {
        match self {
            Self::Dashboard => "Dashboard".into(),
            Self::Home => "Home".into(),
            Self::About => "About".into(),
            Self::NotFound => "404".into(),
            Self::Jobstats => "Jobstats".into(),
        }
    }
}

impl From<Url> for Page {
    fn from(url: Url) -> Self {
        match url.path.first().map(String::as_str) {
            None | Some("") => Self::Home,
            Some("dashboard") => Self::Dashboard,
            Some("about") => Self::About,
            Some("jobstats") => Self::Jobstats,
            _ => Self::NotFound,
        }
    }
}

pub fn register_eventsource_handle<T, F>(
    es_cb_setter: fn(&EventSource, Option<&Function>),
    msg: F,
    ws: &EventSource,
    orders: &mut impl Orders<Msg>,
) where
    T: wasm_bindgen::convert::FromWasmAbi + 'static,
    F: Fn(T) -> Msg + 'static,
{
    let (app, msg_mapper) = (orders.clone_app(), orders.msg_mapper());

    let closure = Closure::new(move |data| {
        app.update(msg_mapper(msg(data)));
    });

    es_cb_setter(ws, Some(closure.as_ref().unchecked_ref()));
    closure.forget();
}

// ------ ------
//     Init
// ------ ------

pub fn init(url: Url, orders: &mut impl Orders<Msg>) -> Init<Model> {
    // @TODO: Seed can't hydrate prerendered html (yet).
    // https://github.com/David-OConnor/seed/issues/223
    if let Some(mount_point_element) = document().get_element_by_id("app") {
        mount_point_element.set_inner_html("");
    }

    let es = EventSource::new("https://localhost:7444/messaging").unwrap();

    register_eventsource_handle(
        EventSource::set_onopen,
        Msg::EventSourceConnect,
        &es,
        orders,
    );

    register_eventsource_handle(
        EventSource::set_onmessage,
        Msg::EventSourceMessage,
        &es,
        orders,
    );

    register_eventsource_handle(
        EventSource::set_onerror,
        Msg::EventSourceError,
        &es,
        orders,
    );

    orders.send_msg(Msg::UpdatePageTitle);

    Init::new(Model {
        page: url.into(),
        menu_visibility: Visible,
        in_prerendering: is_in_prerendering(),
        config_menu_state: WatchState::default(),
        track_slider: false,
        side_width_percentage: 20f32,
        records: warp_drive::Cache::default(),
        locks: HashMap::new(),
    })
}

fn is_in_prerendering() -> bool {
    let user_agent =
        window().navigator().user_agent().expect("cannot get user agent");

    user_agent == USER_AGENT_FOR_PRERENDERING
}

// ------ ------
//    Routes
// ------ ------

pub fn routes(url: Url) -> Option<Msg> {
    // Urls which start with `static` are files => treat them as external links.
    if url.path.starts_with(&[STATIC_PATH.into()]) {
        return None;
    }
    Some(Msg::RouteChanged(url))
}

// ------ ------
//    Update
// ------ ------

#[derive(Clone)]
pub enum Msg {
    RouteChanged(Url),
    UpdatePageTitle,
    ToggleMenu,
    ConfigMenuState,
    HideMenu,
    StartSliderTracking,
    StopSliderTracking,
    SliderX(i32, f64),
    EventSourceConnect(JsValue),
    EventSourceMessage(MessageEvent),
    EventSourceError(JsValue),
    Records(warp_drive::Cache),
    RecordChange(warp_drive::RecordChange),
    Locks(warp_drive::Locks),
    WindowClick,
}

pub fn update(msg: Msg, model: &mut Model, orders: &mut impl Orders<Msg>) {
    match msg {
        Msg::RouteChanged(url) => {
            model.page = url.into();
            orders.send_msg(Msg::UpdatePageTitle);
        }
        Msg::UpdatePageTitle => {
            let title = match model.page {
                Page::Home => TITLE_SUFFIX.to_owned(),
                Page::Dashboard => format!("Dashboard - {}", TITLE_SUFFIX),
                Page::Jobstats => format!("Jobstats - {}", TITLE_SUFFIX),
                Page::About => format!("About - {}", TITLE_SUFFIX),
                Page::NotFound => format!("404 - {}", TITLE_SUFFIX),
            };
            document().set_title(&title);
        }
        Msg::EventSourceConnect(_) => {
            log("EventSource connected.");
        }
        Msg::EventSourceMessage(msg) => {
            let txt = msg.data().as_string().unwrap();

            let msg: warp_drive::Message = serde_json::from_str(&txt).unwrap();

            let msg = match msg {
                warp_drive::Message::Locks(locks) => Msg::Locks(locks),
                warp_drive::Message::Records(records) => Msg::Records(records),
                warp_drive::Message::RecordChange(record_change) => {
                    Msg::RecordChange(record_change)
                }
            };

            orders.send_msg(msg);
        }
        Msg::EventSourceError(_) => {
            log("EventSource error.");
        }
        Msg::Records(records) => {
            model.records = records;
        }
        Msg::RecordChange(record_change) => match record_change {
            warp_drive::RecordChange::Update(record) => match record {
                warp_drive::Record::ActiveAlert(x) => {
                    model.records.active_alert.insert(x.id, x);
                }
                warp_drive::Record::Filesystem(x) => {
                    model.records.filesystem.insert(x.id, x);
                }
                warp_drive::Record::Host(x) => {
                    model.records.host.insert(x.id, x);
                }
                warp_drive::Record::ManagedTargetMount(x) => {
                    model.records.managed_target_mount.insert(x.id, x);
                }
                warp_drive::Record::StratagemConfig(x) => {
                    model.records.stratagem_config.insert(x.id, x);
                }
                warp_drive::Record::Target(x) => {
                    model.records.target.insert(x.id, x);
                }
                warp_drive::Record::Volume(x) => {
                    model.records.volume.insert(x.id, x);
                }
                warp_drive::Record::VolumeNode(x) => {
                    model.records.volume_node.insert(x.id, x);
                }
                warp_drive::Record::LnetConfiguration(x) => {
                    model.records.lnet_configuration.insert(x.id, x);
                }
            },
            warp_drive::RecordChange::Delete(record_id) => match record_id {
                warp_drive::RecordId::ActiveAlert(x) => {
                    model.records.active_alert.remove(&x);
                }
                warp_drive::RecordId::Filesystem(x) => {
                    model.records.filesystem.remove(&x);
                }
                warp_drive::RecordId::Host(x) => {
                    model.records.host.remove(&x);
                }
                warp_drive::RecordId::ManagedTargetMount(x) => {
                    model.records.managed_target_mount.remove(&x);
                }
                warp_drive::RecordId::StratagemConfig(x) => {
                    model.records.stratagem_config.remove(&x);
                }
                warp_drive::RecordId::Target(x) => {
                    model.records.target.remove(&x);
                }
                warp_drive::RecordId::Volume(x) => {
                    model.records.volume.remove(&x);
                }
                warp_drive::RecordId::VolumeNode(x) => {
                    model.records.volume_node.remove(&x);
                }
                warp_drive::RecordId::LnetConfiguration(x) => {
                    model.records.lnet_configuration.remove(&x);
                }
            },
        },
        Msg::Locks(locks) => {
            model.locks = locks;
        }
        Msg::ToggleMenu => model.menu_visibility.toggle(),
        Msg::ConfigMenuState => {
            model.config_menu_state.update();
        }
        Msg::HideMenu => {
            model.menu_visibility = Hidden;
        }
        Msg::StartSliderTracking => {
            model.track_slider = true;
        }
        Msg::StopSliderTracking => {
            model.track_slider = false;
        }
        Msg::SliderX(x_position, page_width) => {
            let overlay_width_px = page_width as u32 - SLIDER_WIDTH_PX;

            let x_position = cmp::max(0, x_position) as u32;

            let side_width_percentage: f32 =
                (x_position as f32 / overlay_width_px as f32) * 100_f32;

            model.side_width_percentage =
                if MAX_SIDE_PERCENTAGE <= side_width_percentage {
                    MAX_SIDE_PERCENTAGE
                } else {
                    side_width_percentage
                };
        }
        Msg::WindowClick => {
            if model.config_menu_state.should_update() {
                model.config_menu_state.update();
            }
        }
    }
}

// ------ ------
//     View
// ------ ------

// Notes:
// - \u{00A0} is the non-breaking space
//   - https://codepoints.net/U+00A0
//
// - "▶\u{fe0e}" - \u{fe0e} is the variation selector, it prevents ▶ to change to emoji in some browsers
//   - https://codepoints.net/U+FE0E

pub fn view(model: &Model) -> impl View<Msg> {
    // @TODO: Setup `prerendered` properly once https://github.com/David-OConnor/seed/issues/223 is resolved
    let prerendered = true;

    div![
        class![
            C.fade_in => !prerendered,
            C.min_h_screen,
            C.flex,
            C.flex_col,
            C.select_none => model.track_slider
        ],
        // slider overlay
        if model.track_slider {
            div![
                class![
                    C.w_full,
                    C.h_full,
                    C.fixed,
                    C.top_0,
                    C.cursor_ew_resize,
                ],
                style! { St::ZIndex => 9999 },
                mouse_ev(Ev::MouseMove, |ev| {
                    let target = ev.target().unwrap();
                    let el = seed::to_html_el(&target);

                    let rect = el.get_bounding_client_rect();

                    Msg::SliderX(ev.client_x(), rect.width())
                }),
            ]
        } else {
            empty![]
        },
        page::partial::header::view(model).els(),
        // panel container
        div![
            class![
                C.flex,
                C.flex_wrap,
                C.flex_col,
                C.lg__flex_row,
                C.flex_grow
            ],
            // side panel
            div![
                class![
                    C.flex_grow_0,
                    C.flex_shrink_0,
                    C.overflow_x_hidden,
                    C.overflow_y_auto,
                    C.whitespace_no_wrap,
                    C.bg_blue_900,
                    C.border_r_2,
                    C.border_gray_800,
                    C.lg__h_main_content,
                ],
                style! { St::FlexBasis => percent(model.side_width_percentage) },
            ],
            // slider panel
            div![
                class![
                    C.flex_grow_0,
                    C.flex_shrink_0,
                    C.cursor_ew_resize,
                    C.bg_gray_500
                    C.hover__bg_teal_400,
                    C.bg_teal_400 => model.track_slider,
                    C.relative,
                    C.lg__block,
                    C.lg__h_main_content,
                    C.hidden
                ],
                simple_ev(Ev::MouseDown, Msg::StartSliderTracking),
                style! {
                    St::FlexBasis => px(SLIDER_WIDTH_PX),
                },
                div![
                    class![C.absolute, C.rounded],
                    style! {
                        St::BackgroundColor => "inherit",
                        St::Height => px(64),
                        St::Width => px(18),
                        St::Top => "calc(50% - 32px)",
                        St::Left => px(-7.5),
                    }
                ]
            ],
            // main panel
            div![
                class![
                    C.flex,
                    C.flex_col,
                    C.flex_grow,
                    C.flex_shrink_0,
                    C.bg_gray_200,
                    C.lg__w_0,
                    C.lg__h_main_content,
                ],
                // main content
                div![
                    class![
                        C.flex_grow,
                        C.overflow_x_auto,
                        C.overflow_y_auto,
                        C.p_6
                    ],
                    match model.page {
                        Page::Home => page::home::view(&model).els(),
                        Page::Dashboard => page::dashboard::view(&model).els(),
                        Page::About => page::about::view(&model).els(),
                        Page::Jobstats => page::jobstats::view(&model).els(),
                        Page::NotFound => page::not_found::view(&model).els(),
                    },
                ],
                page::partial::footer::view().els(),
            ]
        ],
    ]
}

pub fn asset_path(asset: &str) -> String {
    format!("{}/{}", STATIC_PATH, asset)
}

// ------ ------
// Window Events
// ------ ------

pub fn window_events(model: &Model) -> Vec<Listener<Msg>> {
    let mut xs = vec![simple_ev(Ev::Click, Msg::WindowClick)];

    if model.track_slider {
        xs.push(simple_ev(Ev::MouseUp, Msg::StopSliderTracking));
    }

    xs
}

// ------ ------
//     Start
// ------ ------

#[wasm_bindgen(start)]
pub fn run() {
    log!("Starting app...");

    App::build(init, update, view)
        .routes(routes)
        .window_events(window_events)
        .finish()
        .run();

    log!("App started.");
}
