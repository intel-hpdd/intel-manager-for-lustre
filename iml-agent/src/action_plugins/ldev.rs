// Copyright (c) 2020 DDN. All rights reserved.
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file.

use crate::{agent_error::ImlAgentError, env};
use iml_wire_types::LdevEntry;
use std::collections::BTreeSet;
use tokio::{
    fs::{metadata, read_to_string, File},
    io::AsyncWriteExt,
};

async fn write_to_file(content: String) -> Result<(), ImlAgentError> {
    let ldev_path = env::get_ldev_conf();
    let mut file = File::create(ldev_path).await?;
    file.write_all(content.as_bytes()).await?;

    Ok(())
}

async fn read_ldev_config() -> Result<String, ImlAgentError> {
    let ldev_path = env::get_ldev_conf();

    match metadata(&ldev_path).await {
        Ok(_) => Ok(read_to_string(&ldev_path).await?),
        Err(_) => Ok("".into()),
    }
}

fn parse_entries(ldev_config: String) -> BTreeSet<LdevEntry> {
    ldev_config
        .lines()
        .filter(|x| !x.trim_start().starts_with("#"))
        .map(LdevEntry::from)
        .collect()
}

fn convert(entries: &[LdevEntry]) -> String {
    entries
        .iter()
        .map(|x| x.to_string())
        .collect::<Vec<String>>()
        .join("\n")
}

pub async fn create(entries: Vec<LdevEntry>) -> Result<(), ImlAgentError> {
    if !entries.is_empty() {
        let ldev_config = read_ldev_config().await?;
        let existing_entries = parse_entries(ldev_config);
        let entries_set = entries.iter().cloned().collect::<BTreeSet<LdevEntry>>();

        if existing_entries != entries_set {
            let data = convert(&entries);
            write_to_file(data).await?;
        }

        Ok(())
    } else {
        Err(ImlAgentError::LdevEntriesError(
            "The ldev entries must not be empty.".into(),
        ))
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use iml_wire_types::FsType;

    #[test]
    fn test_ldiskfs_create() -> Result<(), ImlAgentError> {
        let entries = vec![
            LdevEntry {
                primary: "mds1".into(),
                failover: Some("mds2".into()),
                label: "MGS".into(),
                device: "/mnt/mgt".into(),
                fs_type: Some(FsType::Ldiskfs),
            },
            LdevEntry {
                primary: "mds1".into(),
                failover: Some("mds2".into()),
                label: "fs-MDT0000".into(),
                device: "/mnt/mdt0".into(),
                fs_type: Some(FsType::Ldiskfs),
            },
            LdevEntry {
                primary: "mds2".into(),
                failover: Some("mds1".into()),
                label: "fs-MDT0001".into(),
                device: "/mnt/mdt1".into(),
                fs_type: Some(FsType::Ldiskfs),
            },
            LdevEntry {
                primary: "oss1".into(),
                failover: Some("oss2".into()),
                label: "fs-OST0000".into(),
                device: "/mnt/ost0".into(),
                fs_type: Some(FsType::Ldiskfs),
            },
            LdevEntry {
                primary: "oss2".into(),
                failover: Some("oss1".into()),
                label: "fs-OST0001".into(),
                device: "/mnt/ost1".into(),
                fs_type: Some(FsType::Ldiskfs),
            },
            LdevEntry {
                primary: "oss1".into(),
                failover: Some("oss2".into()),
                label: "fs-OST0002".into(),
                device: "/mnt/ost2".into(),
                fs_type: Some(FsType::Ldiskfs),
            },
            LdevEntry {
                primary: "oss2".into(),
                failover: Some("oss1".into()),
                label: "fs-OST0003".into(),
                device: "/mnt/ost3".into(),
                fs_type: Some(FsType::Ldiskfs),
            },
        ]
        .into_iter()
        .collect::<Vec<LdevEntry>>();

        let data = convert(&entries);
        insta::assert_snapshot!(data);

        Ok(())
    }

    #[test]
    fn test_zfs_create() -> Result<(), ImlAgentError> {
        let entries = vec![
            LdevEntry {
                primary: "mds1".into(),
                failover: Some("mds2".into()),
                label: "MGS".into(),
                device: "zfs:mdt0/mdt0".into(),
                fs_type: Some(FsType::Zfs),
            },
            LdevEntry {
                primary: "mds1".into(),
                failover: Some("mds2".into()),
                label: "zfsmo-MDT0000".into(),
                device: "zfs:mdt0/mdt0".into(),
                fs_type: Some(FsType::Zfs),
            },
            LdevEntry {
                primary: "mds2".into(),
                failover: Some("mds1".into()),
                label: "zfsmo-MDT0001".into(),
                device: "zfs:mdt1/mdt1".into(),
                fs_type: Some(FsType::Zfs),
            },
            LdevEntry {
                primary: "oss1".into(),
                failover: Some("oss2".into()),
                label: "zfsmo-OST0000".into(),
                device: "zfs:ost0/ost0".into(),
                fs_type: Some(FsType::Zfs),
            },
            LdevEntry {
                primary: "oss1".into(),
                failover: Some("oss2".into()),
                label: "zfsmo-OST0001".into(),
                device: "zfs:ost1/ost1".into(),
                fs_type: Some(FsType::Zfs),
            },
            LdevEntry {
                primary: "oss1".into(),
                failover: Some("oss2".into()),
                label: "zfsmo-OST0002".into(),
                device: "zfs:ost2/ost2".into(),
                fs_type: Some(FsType::Zfs),
            },
            LdevEntry {
                primary: "oss1".into(),
                failover: Some("oss2".into()),
                label: "zfsmo-OST0003".into(),
                device: "zfs:ost3/ost3".into(),
                fs_type: Some(FsType::Zfs),
            },
            LdevEntry {
                primary: "oss1".into(),
                failover: Some("oss2".into()),
                label: "zfsmo-OST0004".into(),
                device: "zfs:ost4/ost4".into(),
                fs_type: Some(FsType::Zfs),
            },
            LdevEntry {
                primary: "oss1".into(),
                failover: Some("oss2".into()),
                label: "zfsmo-OST0005".into(),
                device: "zfs:ost5/ost5".into(),
                fs_type: Some(FsType::Zfs),
            },
            LdevEntry {
                primary: "oss1".into(),
                failover: Some("oss2".into()),
                label: "zfsmo-OST0006".into(),
                device: "zfs:ost6/ost6".into(),
                fs_type: Some(FsType::Zfs),
            },
            LdevEntry {
                primary: "oss1".into(),
                failover: Some("oss2".into()),
                label: "zfsmo-OST0007".into(),
                device: "zfs:ost7/ost7".into(),
                fs_type: Some(FsType::Zfs),
            },
            LdevEntry {
                primary: "oss1".into(),
                failover: Some("oss2".into()),
                label: "zfsmo-OST0008".into(),
                device: "zfs:ost8/ost8".into(),
                fs_type: Some(FsType::Zfs),
            },
            LdevEntry {
                primary: "oss1".into(),
                failover: Some("oss2".into()),
                label: "zfsmo-OST0009".into(),
                device: "zfs:ost9/ost9".into(),
                fs_type: Some(FsType::Zfs),
            },
            LdevEntry {
                primary: "oss2".into(),
                failover: Some("oss1".into()),
                label: "zfsmo-OST000a".into(),
                device: "zfs:ost10/ost10".into(),
                fs_type: Some(FsType::Zfs),
            },
            LdevEntry {
                primary: "oss2".into(),
                failover: Some("oss1".into()),
                label: "zfsmo-OST000b".into(),
                device: "zfs:ost11/ost11".into(),
                fs_type: Some(FsType::Zfs),
            },
            LdevEntry {
                primary: "oss2".into(),
                failover: Some("oss1".into()),
                label: "zfsmo-OST000c".into(),
                device: "zfs:ost12/ost12".into(),
                fs_type: Some(FsType::Zfs),
            },
            LdevEntry {
                primary: "oss2".into(),
                failover: Some("oss1".into()),
                label: "zfsmo-OST000d".into(),
                device: "zfs:ost13/ost13".into(),
                fs_type: Some(FsType::Zfs),
            },
            LdevEntry {
                primary: "oss2".into(),
                failover: Some("oss1".into()),
                label: "zfsmo-OST000e".into(),
                device: "zfs:ost14/ost14".into(),
                fs_type: Some(FsType::Zfs),
            },
            LdevEntry {
                primary: "oss2".into(),
                failover: Some("oss1".into()),
                label: "zfsmo-OST000f".into(),
                device: "zfs:ost15/ost15".into(),
                fs_type: Some(FsType::Zfs),
            },
            LdevEntry {
                primary: "oss2".into(),
                failover: Some("oss1".into()),
                label: "zfsmo-OST0010".into(),
                device: "zfs:ost16/ost16".into(),
                fs_type: Some(FsType::Zfs),
            },
            LdevEntry {
                primary: "oss2".into(),
                failover: Some("oss1".into()),
                label: "zfsmo-OST0011".into(),
                device: "zfs:ost17/ost17".into(),
                fs_type: Some(FsType::Zfs),
            },
            LdevEntry {
                primary: "oss2".into(),
                failover: Some("oss1".into()),
                label: "zfsmo-OST0012".into(),
                device: "zfs:ost18/ost18".into(),
                fs_type: Some(FsType::Zfs),
            },
            LdevEntry {
                primary: "oss2".into(),
                failover: Some("oss1".into()),
                label: "zfsmo-OST0013".into(),
                device: "zfs:ost19/ost19".into(),
                fs_type: Some(FsType::Zfs),
            },
        ]
        .into_iter()
        .collect::<Vec<LdevEntry>>();

        let data = convert(&entries);
        insta::assert_snapshot!(data);

        Ok(())
    }

    #[test]
    fn test_config_without_ha() -> Result<(), ImlAgentError> {
        let entries = vec![
            LdevEntry {
                primary: "mds1".into(),
                failover: None,
                label: "MGS".into(),
                device: "zfs:mdt0/mdt0".into(),
                fs_type: Some(FsType::Zfs),
            },
            LdevEntry {
                primary: "mds1".into(),
                failover: Some("mds2".into()),
                label: "zfsmo-MDT0000".into(),
                device: "zfs:mdt0/mdt0".into(),
                fs_type: Some(FsType::Zfs),
            },
        ]
        .into_iter()
        .collect::<Vec<LdevEntry>>();

        let data = convert(&entries);
        insta::assert_snapshot!(data);

        Ok(())
    }

    #[test]
    fn test_parsing_commented_data() -> Result<(), ImlAgentError> {
        let content: String = r#"# example /etc/ldev.conf
#
#local  foreign/-  label       [md|zfs:]device-path   [journal-path]/- [raidtab]
#
#zeno-mds1 - zeno-MDT0000 zfs:lustre-zeno-mds1/mdt1
#
#zeno1 zeno5 zeno-OST0000 zfs:lustre-zeno1/ost1
#zeno2 zeno6 zeno-OST0001 zfs:lustre-zeno2/ost1
#zeno3 zeno7 zeno-OST0002 zfs:lustre-zeno3/ost1
#zeno4 zeno8 zeno-OST0003 zfs:lustre-zeno4/ost1
#zeno5 zeno1 zeno-OST0004 zfs:lustre-zeno5/ost1
#zeno6 zeno2 zeno-OST0005 zfs:lustre-zeno6/ost1
#zeno7 zeno3 zeno-OST0006 zfs:lus tre-zeno7/ost1
#zeno8 zeno4 zeno-OST0007 zfs:lustre-zeno8/ost1"#
            .into();

        let data: BTreeSet<LdevEntry> = parse_entries(content);

        assert!(data.is_empty());

        Ok(())
    }

    #[test]
    fn test_parsing_data() -> Result<(), ImlAgentError> {
        let content: String = r#"#zeno-mds1 - zeno-MDT0000 zfs:lustre-zeno-mds1/mdt1
# Random comment
zeno1 zeno5 zeno-OST0000 zfs:lustre-zeno1/ost1
zeno2 - zeno-OST0001 zfs:lustre-zeno2/ost1
zeno3 zeno7 zeno-OST0002 zfs:lustre-zeno3/ost1
zeno4 zeno8 zeno-OST0003 zfs:lustre-zeno4/ost1
zeno5 zeno1 zeno-OST0004 zfs:lustre-zeno5/ost1
zeno6 - zeno-OST0005 zfs:lustre-zeno6/ost1
zeno7 zeno3 zeno-OST0006 zfs:lustre-zeno7/ost1
zeno8 zeno4 zeno-OST0007 zfs:lustre-zeno8/ost1"#
            .into();

        let data: BTreeSet<LdevEntry> = parse_entries(content);

        insta::assert_debug_snapshot!(data);

        Ok(())
    }
}
