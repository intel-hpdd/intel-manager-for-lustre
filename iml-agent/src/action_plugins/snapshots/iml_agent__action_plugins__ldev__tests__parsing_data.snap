---
source: iml-agent/src/action_plugins/ldev.rs
expression: data
---
{
    LdevEntry {
        primary: "zeno1",
        failover: Some(
            "zeno5",
        ),
        label: "zeno-OST0000",
        device: "zfs:lustre-zeno1/ost1",
        fs_type: Some(
            Zfs,
        ),
    },
    LdevEntry {
        primary: "zeno2",
        failover: None,
        label: "zeno-OST0001",
        device: "zfs:lustre-zeno2/ost1",
        fs_type: Some(
            Zfs,
        ),
    },
    LdevEntry {
        primary: "zeno3",
        failover: Some(
            "zeno7",
        ),
        label: "zeno-OST0002",
        device: "zfs:lustre-zeno3/ost1",
        fs_type: Some(
            Zfs,
        ),
    },
    LdevEntry {
        primary: "zeno4",
        failover: Some(
            "zeno8",
        ),
        label: "zeno-OST0003",
        device: "zfs:lustre-zeno4/ost1",
        fs_type: Some(
            Zfs,
        ),
    },
    LdevEntry {
        primary: "zeno5",
        failover: Some(
            "zeno1",
        ),
        label: "zeno-OST0004",
        device: "zfs:lustre-zeno5/ost1",
        fs_type: Some(
            Zfs,
        ),
    },
    LdevEntry {
        primary: "zeno6",
        failover: None,
        label: "zeno-OST0005",
        device: "zfs:lustre-zeno6/ost1",
        fs_type: Some(
            Zfs,
        ),
    },
    LdevEntry {
        primary: "zeno7",
        failover: Some(
            "zeno3",
        ),
        label: "zeno-OST0006",
        device: "zfs:lustre-zeno7/ost1",
        fs_type: Some(
            Zfs,
        ),
    },
    LdevEntry {
        primary: "zeno8",
        failover: Some(
            "zeno4",
        ),
        label: "zeno-OST0007",
        device: "zfs:lustre-zeno8/ost1",
        fs_type: Some(
            Zfs,
        ),
    },
}
